NewHaven/src/VGMap.cpp

```diff
@@ -215,77 +229,72 @@ void VGMap::resetVisited() {
 }
 
 void VGMap::flipMapBuilding(int position) {
-    (*village_board)[position].flipBuilding();
+    if(*(*village_board)[position].isPlayed)
+        (*village_board)[position].building->flipCard();
 }
 
 void VGMap::setBuilding(int position, Building *building) {
-    (*village_board)[position].setBuilding(building);
+    (*village_board)[position].building = new Building(*building);
+    *(*village_board)[position].isPlayed = true;
+}
+
+VGMap::VGMap(const VGMap &map) {
+    village_board = new C_Graph(*map.village_board);
 }
-Circle::Circle(){
 
+Circle::Circle(){
+    row = nullptr;
+    column = nullptr;
+    vCost = nullptr;
+    building = nullptr;
+    position = nullptr;
+    isVisited = new bool(false);
+    isPlayed = new bool(false);
 }
 
 Circle::Circle(const Circle &circle){
     if(circle.building != nullptr){
-        building = new Building(*circle.building);
+         building = new Building(*circle.building);
     } else
         building = nullptr;
     row = new int(*circle.row);
+    position = new int(*circle.position);
     column = new int(*circle.column);
     vCost = new int(*circle.vCost);
     isVisited = new bool(*circle.isVisited);
     isPlayed = new bool(*circle.isPlayed);
 }
 
-Circle::~Circle() = default;
-int Circle::getRow() const{
-    return *row;
-}
-
-int Circle::getColumn() const{
-    return *column;
+Circle::~Circle(){
+    delete row;
+    delete column;
+    delete vCost;
+    delete isVisited;
+    delete position;
+    delete isPlayed;
+    delete building;
 }
 
-int Circle::getPosition() const{
-    return *position;
-}
-
-int Circle::getVCost() const{
-    return *vCost;
-}
-
-Building* Circle::getBuilding() const{
-    return building;
-}
-
-void Circle::setCol(int col) {
-    column = new int(col);
-}
-
-void Circle::setRow(int row) {
-    this->row = new int(row);
-}
-
-void Circle::setVCost(int cost) {
-    vCost = new int(cost);
-}
-
-void Circle::setBuilding(Building * building) {
-    if(!*isPlayed) {
-        this->building = new Building(*building);
-        *this->isPlayed = true;
-        return;
+// override assignment operator
+Circle & Circle::operator=(const Circle &circle){
+    if (this == &circle)
+        return *this;
+    else {
+        delete row;
+        delete column;
+        delete vCost;
+        delete isVisited;
+        delete isPlayed;
+        delete building;
+        row = new int(*circle.row);
+        column = new int(*circle.column);
+        vCost = new int(*circle.vCost);
+        isVisited = new bool(*circle.isVisited);
+        isPlayed = new bool(*circle.isPlayed);
+        // for some reason copy constructor of building is causing a segFault here
+        building = new Building();
+        *building = *circle.building;
+        position = new int(*circle.position);
```

Comment:
`building = new Building(*circle.building);`
In one line. But what if you're copying from a circle that has building as nullptr?
You're not making any check.

Reply:
even if a building is null and is passed as an argument then nothing happens.